internal/querycoord/query_coord.go

```diff
@@ -507,29 +507,27 @@ func (qc *QueryCoord) watchHandoffSegmentLoop() {
 			return
 		case resp := <-watchChan:
 			for _, event := range resp.Events {
+				if event.Type != mvccpb.PUT {
+					continue
+				}
+
 				segmentInfo := &querypb.SegmentInfo{}
-				err := proto.Unmarshal(event.Kv.Value, segmentInfo)
-				if err != nil {
-					log.Error("watchHandoffSegmentLoop: unmarshal failed", zap.Any("error", err.Error()))
+				if err := proto.Unmarshal(event.Kv.Value, segmentInfo); err != nil {
+					log.Error("watchHandoffSegmentLoop: unmarshal failed", zap.Error(err))
 					continue
 				}
-				switch event.Type {
-				case mvccpb.PUT:
-					validHandoffReq, _ := qc.indexChecker.verifyHandoffReqValid(segmentInfo)
-					if Params.QueryCoordCfg.AutoHandoff && validHandoffReq {
-						qc.indexChecker.enqueueHandoffReq(segmentInfo)
-						log.Info("watchHandoffSegmentLoop: enqueue a handoff request to index checker", zap.Any("segment info", segmentInfo))
-					} else {
-						log.Info("watchHandoffSegmentLoop: collection/partition has not been loaded or autoHandoff equal to false, remove req from etcd", zap.Any("segmentInfo", segmentInfo))
-						buildQuerySegmentPath := fmt.Sprintf("%s/%d/%d/%d", handoffSegmentPrefix, segmentInfo.CollectionID, segmentInfo.PartitionID, segmentInfo.SegmentID)
-						err = qc.kvClient.Remove(buildQuerySegmentPath)
-						if err != nil {
-							log.Error("watchHandoffSegmentLoop: remove handoff segment from etcd failed", zap.Error(err))
-							panic(err)
-						}
+
+				validHandoffReq, _ := qc.indexChecker.verifyHandoffReqValid(segmentInfo)
+				if Params.QueryCoordCfg.AutoHandoff && validHandoffReq {
+					go qc.indexChecker.handleHandoffRequest(segmentInfo)
```

Comment:
question: Doing it parallel here, will cause the later handoff tasks to enter the queue first?

Reply:
Yes. But it does not harm safety if we make sure that these tasks are not correlated.
This is due to the reference lock that the pending handoff segment will not be compacted.